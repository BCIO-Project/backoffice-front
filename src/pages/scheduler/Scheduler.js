import React, { Component } from 'react';
import { Redirect } from 'react-router';
import {
    Button,
    Form
} from 'semantic-ui-react';
import { Calendar, momentLocalizer } from 'react-big-calendar';
import 'react-big-calendar/lib/css/react-big-calendar.css';
import moment from 'moment';

import MainHeader from '../../components/header/MainHeader';
import SideBar from './../home/components/SideBar';
import * as pagesService from './../../services/pages.service';
import * as campaignService from '../../services/campaign.service';
import './Scheduler.scss';

const localizer = momentLocalizer(moment);

const colors = ['#FCF2D1', '#DAF6EC', '#FDDBDF', '#D8E7F9', '#E9DDEE', '#ECF1C5', '#BEDAC8', '#F1B3D0', '#C5E1F7', '#DDB6E6', '#F9E0BB', '#D9EAC8', '#F1B3B8', '#D9DBF0', '#FACABF', '#C9CAE9', '#C1E1DC', '#C9EBD8', '#E5E2DE', '#D0D0D0']


export default class Scheduler extends Component {
    constructor(props) {
        super(props);
        this.menuRef = React.createRef()
        this.toggleRef = React.createRef()
    }
    state = {
        increaseSidebar: false,
        page: '',
        optionPages: [],
        optionPositions: [],
        originalCampaigns: [],
        filteredCampaigns: [],
        events: [],
        toggleLeyend: false,
        paginationPage: 1,
    }

    /**
     * life cycle React component
     */
    componentDidMount() {
        // search portada campaigns when page is loaded
        try {
            this.getCampaigns()
            this.getPages()
        } catch (error) {
            console.log(error)
        }
    }

    /**
   * life cycle React component
   * @param {object} prevProps 
   * @param {object} prevState 
   */
    componentDidUpdate(prevProps, prevState) {
        // get positions when page changes
        if (prevState.page !== this.state.page && this.state.page !== '') {
            this.getPositionsByPage()
        }
        // filter campaings by page when positions changes
        if (prevState.optionPositions !== this.state.optionPositions && this.state.optionPositions.length && this.state.originalCampaigns.length) {
            this.filterCampaignsbyPage()
        }
        // transform filteredCampaigns arr into events arr for scheduler
        if (prevState.filteredCampaigns !== this.state.filteredCampaigns) {
            this.transformFilteredCampaignsToEventsArr(this.state.filteredCampaigns)
        }
    }

    /**
     * fetch all campaigns
     * @memberof Scheduler
     */
    getCampaigns = () => {
        // transform filters data's format and send it to API
        campaignService.get('', '', `?page=${this.state.paginationPage}`)
            .then(data => {
                this.setState((prevState)=>{
                    // store campaings in component's state, determining if there's need to load another page
                    return {
                      originalCampaigns: this.state.paginationPage === 1 ? data : [...prevState.originalCampaigns, ...data],
                      hasMore: data.length === 15 ? true : false,  
                      paginationPage: this.state.paginationPage + 1
                    }
                  }, () => {
                      if(this.state.hasMore) {
                          this.getCampaigns()
                      }
                  })
            })
    }

    /**
     * fetch all pages
     * @memberof Scheduler
     */
    getPages = () => {
        pagesService.getList()
            .then((data) => {
                this.setState({
                    optionPages: data
                        .map(data => ({
                            key: data.id,
                            value: data.name,
                            text: data.name,
                            id: data.id
                        }))
                }, this.getDefaultPageID(data))
            })
    }

    /**
     * get lower page ID to set default page
     * @memberof Scheduler
     */
    getDefaultPageID = (pages) => {
        const pageIDs = pages.map(page => page.id)
        const orderArr = pageIDs.sort((a, b) => a - b)
        this.setState({
            page: orderArr[0]
        })
    }

    /**
     * fetch positions by page
     * @memberof Scheduler
     */
    getPositionsByPage = () => {
        pagesService.getList(`${this.state.page}/positions`)
            .then((data) => {
                this.setState({
                    optionPositions: this.getPositionsWithColors(data)
                })
            })
    }

    /**
     * filter campaigns by page selected and store them in component's state
     * @memberof Scheduler
     */
    filterCampaignsbyPage = () => {
        // eslint-disable-next-line
        const filteredCampaigns = this.state.originalCampaigns.filter(campaign => campaign.page.id == this.state.page)
        this.setState({
            filteredCampaigns: filteredCampaigns.map(campaign => ({ ...campaign, color: this.findColor(campaign.position.id) }))
        })
    }

    /**
     * get color from position item that matched in campaign's positionId and positionId comparision
     * @param {number} campaignPositionId 
     * @memberof Scheduler
     */
    findColor = (campaignPositionId) => {
        const index = this.state.optionPositions.findIndex(item => item.id === campaignPositionId)
        return this.state.optionPositions[index].color
    }

    /**
     * asign color to position
     * @param {array} positions 
     * @memberof Scheduler
     */
    getPositionsWithColors = (positions) => {
        // make colors array copy
        let colorCopy = [...colors];
        return positions
            .map((data, index) => {
                // if colors array doesn't have any values make another copy
                if (!colorCopy.length) {
                    colorCopy = [...colors];
                }
                return {
                    key: data.id,
                    value: data.name,
                    text: data.name,
                    id: data.id,
                    // get first value from colors copy array
                    color: colorCopy.shift()
                }
            });
    }

    /**
     * transform api campaign object to calendar object 
     * @param {object} apiObj - campaign's object from api
     * @memberof Scheduler
     */
    transformApiCampaignObjToCalendarObj = (apiObj) => {
        return {
            id: apiObj.id,
            title: apiObj.name,
            start: new Date(apiObj.from),
            end: new Date(apiObj.to),
            color: apiObj.color,
            status: apiObj.status,
            slot: apiObj.position.name
        }
    }

    /**
     * get events array and store it in component's state
     * @param {array} arr - filtered campaigns array by page
     * @memberof Scheduler
     */
    transformFilteredCampaignsToEventsArr = (arr) => {
        const events = arr.map(item => this.transformApiCampaignObjToCalendarObj(item))
        this.setState({
            events: events
        })
    }    

    /**
   * recover page, position and title values and save them in component's state
   * @memberof Scheduler
   * @param {object} event - The onChange input's event. 
   */
    handleChange = (e) => {
        const updateState = { [e.target.name]: e.target.value }
        this.setState(updateState)
    }

    /**
     * hide and show leyend onclick
     * @memberof Scheduler
     */
    handleLeyendVisibility = () => {
        this.setState({toggleLeyend: !this.state.toggleLeyend})
    }

    /**
     * modify scheduler tooltip text 
     * @param {object} event - scheduler's event object from array passed
     * @memberof Scheduler
     */
    handleTooltip(event) {
        return `${event.title} | ${event.slot} | ${event.status}`
    }

    /**
     * modify scheduler tooltip text 
     * @param {object} event - scheduler's event object from array passed
     * @memberof Scheduler
     */
    onSelectCampaign = (event) => {
        this.setState({ 
            redirect: true, 
            campaignId: event.id 
        })
    }

    /**
     * event styles
     * @param {object} event - scheduler's event object from array passed
     * @memberof Scheduler
     */
    eventStyleGetter(event) {
        const backgroundColor = event.color;
        const border = event.status === 'DRAFT' ? '2px dotted #88868B' : 'none';
        const style = {
            background: backgroundColor,
            borderRadius: '0px',
            color: 'black',
            border: border,
            display: 'block',
            fontFamily: "'Quattrocento Sans', 'sans-serif'",
            fontWeight: '700',
        };
        return {
            style: style
        };
    }

    /**
     * day styles
     * @memberof Scheduler
     */
    dayStyleGetter() {
        const style = {
            borderLeft: '0.1px solid #b5bffd',
        };
        return {
            style: style
        };
    }
    render() {
        const { increaseSidebar, showNotifications, campaignId, redirect } = this.state
        if (redirect) {
            return <Redirect push to={`home/${campaignId}`} />;
        }
        return (
            <section className="home_page">
                <aside className={`home_page_sidebar ${increaseSidebar ? 'open' : ''} ${showNotifications ? 'notifications' : ''}`} ref={this.menuRef}>
                    <SideBar
                        increaseSidebar={increaseSidebar}
                        schedulerView={true}
                    />
                </aside>
                <article className='scheduler_page_container'>
                    <MainHeader className="home" logoHidden={true}></MainHeader>
                    <div className='scheduler_manager'>
                        <div className='scheduler_leyend_container'>
                            <Button className='leyend_btn' onClick={this.handleLeyendVisibility}>Legend</Button>
                            <div className={this.state.toggleLeyend ? 'leyend_content' : 'hidden'}>
                                {this.state.optionPositions
                                    .map(position =>
                                        <p key={position.id}><span className='leyend_color' style={{ backgroundColor: position.color }}></span>{position.value}</p>
                                    )}
                            </div>
                        </div>
                        <div>
                            <Form.Field>
                                <select name="page" id="page" onChange={this.handleChange} className='bcio' value={this.state.page}>
                                    <option value='' className='placeholder'>Filter by page</option>
                                    {this.state.optionPages
                                        .map(option => <option value={option.id} key={option.key}>{option.value}</option>)}
                                </select>
                            </Form.Field>
                        </div>
                    </div>
                    <div style={{ height: '620pt' }} className='calendar_component'>
                        <Calendar
                            events={this.state.events}
                            startAccessor="start"
                            endAccessor="end"
                            defaultDate={moment().toDate()}
                            view='month'
                            onView={() => console.log('month view')}
                            views={['month']}
                            localizer={localizer}
                            onSelectEvent={this.onSelectCampaign}
                            eventPropGetter={(this.eventStyleGetter)}
                            tooltipAccessor={this.handleTooltip}
                            dayPropGetter={(this.dayStyleGetter)}
                            popup={true}
                        />
                    </div>
                </article>
            </section>
        )
    }
}