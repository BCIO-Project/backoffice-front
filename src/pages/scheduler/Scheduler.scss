@import '../../scss/vars';

.scheduler_page_container {
    width: calc(100% - 80px);
    background-color: $color-background;
    position: relative;
    @media screen and (max-width: $tablet-portrait-width){
        width: 100%;
    }
    .main-header {
        box-shadow: none;
    }
    .scheduler_manager {
        position: absolute;
        right: 40px;
        display: flex;
        @media screen and (max-width: $tablet-portrait-width){
            bottom: 0;
            background-color: rgba($color-primary-dark-900, .6);
            height: 70px;
            z-index: 20;
            width: 100%;
            align-items: center;
            right: 0;
            justify-content: space-between;
            padding: 0 10px;
        }
        .scheduler_leyend_container {
            position: relative;
        }
        .leyend_color {
            display: inline-block;
            width: 14px;
            height: 14px;
            border-radius: 5px;
            margin-right: 5px;
        }
        .ui.button.leyend_btn {
            background-color: $color-primary;
            border: 1px solid $color-active-100;
            color: $color-active-900;
            margin-right: 10px;
            height: 40px;
            outline: none;
            border-radius: 30px;
            @media screen and (max-width: $tablet-portrait-width){
                font-size: 11px;
                width: 80px;
            }
        }
        .leyend_content {
            z-index: 10;
            position: absolute;
            background-color: $color-primary;
            padding: 15px 0 20px 15px;
            border: 1px solid $color-active-300;
            width: 186px;
            border-radius: 5px;
            text-transform: uppercase;
            right: 10px;
            margin-top: 7px;
            @media screen and (max-width: $tablet-portrait-width){
                bottom: 55px;
                left: 0;
            }
        }
    }
    .bcio {
        height: 38px;
        background-color: $color-primary;
        font-family: Lato,'Helvetica Neue', Arial, Helvetica, sans-serif;
        border-radius: 3px;
        border: 1px solid $color-primary-light-300;
        width: 200px;
        color: $color-active-900;
        @media screen and (max-width: $tablet-portrait-width){
            font-size: 11px;
            width: 140px;
        }
    }
    option:not(:first-of-type) {
        color: $color-primary-light-900;
    }
    .field {
        .bcio {
            padding-left: 5px;
        }
    }
}
.calendar_component {
    margin: 0 40px 0 30px;
    @media screen and (max-width: $tablet-portrait-width){
        margin: 0 10px;
        height: 70vh !important;
    }
    @media screen and (max-width: $small-mobile-portrait-width){
        margin: 0 10px;
        height: 68vh !important;
    }
    .rbc-toolbar-label {
      color: $color-primary-dark-900;	
      font-family: 'Cabin', sans-serif;	
      font-size: 32px;	
      font-weight: bold;
      margin: 15px 0 5px 0;
        @media screen and (max-width: $tablet-portrait-width){
            font-size: 18px;
            text-align: end;
            padding-right: 5px;
        }
    }
    .rbc-toolbar {
      align-items: flex-end;
      width: 85%;
      margin-bottom: 18px;
        @media screen and (max-width: $tablet-portrait-width){
            flex-wrap: nowrap;
            justify-content: space-between;
            width: 100%;
        }
        button {
            border: 1px solid $color-active-300;
            border-radius: 30px;
            height: 40px;
            width: 78px;
            outline: none;
            color: $color-active-900;
            font-family: Lato,'Helvetica Neue', Arial, Helvetica, sans-serif;
            font-size: 13px;
            font-weight: bold;
            &:hover {
                background-color: $color-active-300;
                color: $color-primary;
            }
            &:active {
                background-color: $color-active-300;
                color: $color-primary;
            }
            &:first-child:focus {
                background-color: $color-active-300;
                color: $color-primary;
            }
            @media screen and (max-width: $tablet-portrait-width){
                font-size: 11px;
                height: 33px;
                width: 55px;
            }
        }
    }
    .rbc-today {
      background-color: $color-active-300;
    }
    .rbc-month-row + .rbc-month-row {
      border-top: 1px solid $color-active-300;
      height: 178px;
    }
    .rbc-month-view {
      border: 1px solid $color-active-300;
      border-radius: 7px;
      background-color: $color-primary;
      overflow: hidden;
    }
    .rbc-date-cell {
      color: $color-active-900;
      font-size: 14px;
      padding: 18px 15px 15px;
      @media screen and (max-width: $tablet-portrait-width){
        font-size: 10px;
        padding: 5px;
        }
      &.rbc-now {
        color: $color-primary;
        position: relative;
        z-index: 5;
        a {
          display: inline-block;
          width: 30px;
          height: 30px;
          border-radius: 50%;
          background-color: $color-active-900;
          text-align: center;
          padding-top: 5px;
          @media screen and (max-width: $tablet-portrait-width){
              width: 20px;
              height: 20px;
              padding-top: 0;
          } 
        }
      }
      a {
        cursor: default;
      }
    }
    .rbc-off-range-bg {
      background-color: $color-background;
    }
    .rbc-month-header {
      text-transform: uppercase;
      font-size: 13px;
      .rbc-header {
        padding: 4px 0;
        border-bottom: 1px solid $color-active-300;
      }
    }
    .rbc-header + .rbc-header {
      border-left: 1px solid $color-active-300;
    }
    .rbc-event {
        font-size: 14px;
        height: 25px;
        @media screen and (max-width: $tablet-portrait-width){
            font-size: 11px;
            height: 16px;
        }
    }
    .rbc-event-content {
        @media screen and (max-width: $tablet-portrait-width){
            transform: translate(0px, -5px);
        }
    }
    .rbc-btn-group {
        display: flex;
        button {
            background-color: $color-primary;
        }
    }
    .rbc-btn-group > button:not(:first-child):not(:last-child) {
        border-radius: 30px;
        border-top-right-radius: 0;
        border-bottom-right-radius: 0;
        order: -1;
        border-right: none;
    }
    .rbc-btn-group > button:first-child:not(:last-child) {
        border-radius: 0;
    }
  }
  