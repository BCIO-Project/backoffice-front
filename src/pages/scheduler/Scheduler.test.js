import React from 'react';
import {shallow} from 'enzyme';
import Scheduler from './Scheduler';

const defaultProps = {
    match: {
        params: {}
    }
}

const optionPositions = [
    {
        color: "#FCF2D1",
        id: 34,
        key: 34,
        text: "Premium",
        value: "Premium"
    }
]

const apiObject = {
    createdAt: "2020-02-14T14:48:05.925Z",
    deletedAt: null,
    from: "2020-02-04T23:00:00.000Z",
    id: 106,
    lastNotificationDate: null,
    name: "Tests01",
    page: {id: 28, name: "El PaÃ­s - PORTADA", slug: "el-pais-portada", createdAt: "2019-10-10T13:52:54.827Z"},
    position: {id: 34, name: "Premium", createdAt: "2019-10-11T11:30:25.552Z", updatedAt: "2019-10-11T11:30:25.552Z"},
    status: "DRAFT",
    to: "2020-02-13T23:00:00.000Z",
    updatedAt: "2020-02-14T14:48:05.940Z",
    color: "#FCF2D1",
}

const e = {
    preventDefault() {},
    target: { value: 1, name: 'page' }
};

describe('Scheduler component', ()=>{
    let scheduler;
    beforeEach(()=>{
        scheduler = shallow(<Scheduler></Scheduler>)
    })
    it('renders correctly', ()=>{
        expect(scheduler).toMatchSnapshot();
    })
    it('should return color string', () => {
        scheduler.setState({ optionPositions: optionPositions })
        const id = 34;
        expect(scheduler.instance().findColor(id)).toStrictEqual('#FCF2D1')
    });
    it('should transform object into scheduler object', () => {
        expect(scheduler.instance().transformApiCampaignObjToCalendarObj(apiObject)).toStrictEqual({
            id: 106,
            title: "Tests01",
            start: new Date("2020-02-04T23:00:00.000Z"),
            end: new Date("2020-02-13T23:00:00.000Z"),
            color: "#FCF2D1",
            status: "DRAFT",
            slot: "Premium"
        })
    });
    test('should call handleLeyendVisibility onClick and show leyend', ()=>{
        const instance = scheduler.instance();
        const handleLeyendVisibilitySpy = jest.spyOn(instance, 'handleLeyendVisibility');
        instance.forceUpdate();
        scheduler.find('.leyend_btn').simulate('click')

        const leyendContainer= scheduler.find('.leyend_content')
        expect(handleLeyendVisibilitySpy).toHaveBeenCalled();
        expect(!leyendContainer.hasClass('hidden'));
    });
    test('should change redirect and campaignId state', ()=>{
        const event = { id: 37 };
        const instance = scheduler.instance()

        instance.onSelectCampaign(event)
        expect(scheduler.state().redirect).toStrictEqual(true)
        expect(scheduler.state().campaignId).toStrictEqual(event.id)
    });
})