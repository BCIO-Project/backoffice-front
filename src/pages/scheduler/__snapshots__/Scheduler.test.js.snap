// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scheduler component renders correctly 1`] = `ShallowWrapper {}`;
